iroh-sync/src/store.rs

```diff
@@ -143,233 +142,243 @@ pub trait Store: std::fmt::Debug + Clone + Send + Sync + 'static {
+            direction: ordering,
+        } = builder;
+
+        Query {
+            kind: QueryKind::Flat(kind),
+            filter_author,
+            filter_key,
+            limit_offset,
+            include_empty,
+            ordering,
+        }
     }
 }
 
-/// A range.
-#[derive(Debug, Serialize, Deserialize, Clone)]
-pub enum Range {
-    /// All entries (no limit)
-    All,
-    /// All entries after offset `x`.
-    From(u64),
-    /// Entries from start to `x` (exclusive)
-    To(u64),
-    /// Entries from start to `x` (inclusive).
-    ToInclusive(u64),
-    /// Entries from start `x` to end `y` (inclusive)
-    Inclusive(u64, u64),
-    /// Entries from start `x` to end `y` (exclusive)
-    Exclusive(u64, u64),
+/// Note: When using the `SingleLatestPerKey` query kind, the key filter is applied *before* the
+/// grouping, the author filter is applied *after* the grouping.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Query {
+    kind: QueryKind,
+    filter_author: AuthorMatcher,
+    filter_key: KeyMatcher,
+    limit_offset: LimitOffset,
+    include_empty: bool,
+    ordering: Direction,
 }
 
-impl Default for Range {
-    fn default() -> Self {
-        Range::All
+impl Query {
+    /// Query all records.
+    pub fn all() -> QueryBuilder<FlatQuery> {
+        Default::default()
+    }
+    /// Query only the latest entry for each key, omitting older entries if the entry was written
+    /// to by multiple authors.
+    pub fn single_latest_per_key() -> QueryBuilder<SingleLatestPerKeyQuery> {
+        Default::default()
     }
-}
 
-impl Range {
-    /// The start of the range.
-    pub fn start(&self) -> u64 {
-        match self {
-            Self::All => 0,
-            Self::From(start) => *start,
-            Self::ToInclusive(_) => 0,
-            Self::Inclusive(start, _) => *start,
-            Self::To(_) => 0,
-            Self::Exclusive(start, _) => *start,
-        }
+    /// Create a [Query::all] query filtered by a single author.
+    pub fn author(author: AuthorId) -> QueryBuilder<FlatQuery> {
+        Self::all().author(author)
     }
 
-    /// The optional end of the range, exclusive.
-    pub fn end(&self) -> Option<u64> {
-        match self {
-            Self::All => None,
-            Self::From(_) => None,
-            Self::ToInclusive(end) => end.checked_add(1),
-            Self::Inclusive(_, end) => end.checked_add(1),
-            Self::To(end) => Some(*end),
-            Self::Exclusive(_, end) => Some(*end),
-        }
+    /// Create a [Query::all] query filtered by a single key.
+    pub fn key(key: impl AsRef<[u8]>) -> QueryBuilder<FlatQuery> {
+        Self::all().key_exact(key)
     }
-}
 
-impl From<std::ops::Range<u64>> for Range {
-    fn from(value: std::ops::Range<u64>) -> Self {
-        Range::Exclusive(value.start, value.end)
+    /// Create a [Query::all] query filtered by a key prefix.
+    pub fn prefix(prefix: impl AsRef<[u8]>) -> QueryBuilder<FlatQuery> {
```

Comment:
maybe use `key_exact` and `key_prefix` for these too?

Reply:
done